lib/interceptor/cache.js

```diff
@@ -0,0 +1,148 @@
+'use strict'
+
+const CacheHandler = require('../handler/cache-handler')
+const MemoryCacheStore = require('../cache/memory-cache-store')
+const CacheRevalidationHandler = require('../handler/cache-revalidation-handler')
+const { UNSAFE_METHODS } = require('../util/cache.js')
+
+const AGE_HEADER = Buffer.from('age')
+
+/**
+ * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions | undefined} globalOpts
+ * @returns {import('../../types/dispatcher.d.ts').default.DispatcherComposeInterceptor}
+ */
+module.exports = globalOpts => {
+  if (!globalOpts) {
+    globalOpts = {}
+  }
+
+  if (globalOpts.store) {
+    for (const fn of ['createReadStream', 'createWriteStream', 'deleteByOrigin']) {
+      if (typeof globalOpts.store[fn] !== 'function') {
+        throw new TypeError(`CacheStore needs a \`${fn}()\` function`)
+      }
+    }
+
+    if (typeof globalOpts.store.isFull !== 'boolean') {
+      throw new TypeError(`CacheStore needs a isFull getter with type boolean, current type: ${typeof globalOpts.store.isFull}`)
+    }
+  } else {
+    globalOpts.store = new MemoryCacheStore()
+  }
+
+  if (globalOpts.methods) {
+    if (!Array.isArray(globalOpts.methods)) {
+      throw new TypeError(`methods needs to be an array, got ${typeof globalOpts.methods}`)
+    }
+
+    if (globalOpts.methods.length === 0) {
+      throw new Error('methods must have at least one method in it')
+    }
+  } else {
+    globalOpts.methods = ['GET']
+  }
+
+  // Safe methods the user wants and unsafe methods
+  const methods = [...globalOpts.methods, ...UNSAFE_METHODS]
+
+  return dispatch => {
+    return (opts, handler) => {
+      if (!opts.origin || !methods.includes(opts.method)) {
+        // Not a method we want to cache or we don't have the origin, skip
+        return dispatch(opts, handler)
+      }
+
+      const stream = globalOpts.store.createReadStream(opts)
+      if (!stream) {
+        // Request isn't cached
+        return dispatch(opts, new CacheHandler(globalOpts, opts, handler))
+      }
+
+      const { value } = stream
```

Comment:
This is weird. Why does stream have a  `value`?  I think you can just put the required state props directly on the stream object.

Reply:
The `data` event doesn't get the `value` since that event is just used for sending the body chunks to the interceptor.

The cache store could send the `data` event something like `{value: {/* cachestore value */}, body: [/* body chunks read */]}`, but that would just take up more unnecessary space in memory for each chunk. We could just only send the `value` property on the first call to read but I'm not sure on how I feel about that